flips/20220203-capability-controllers.md

```diff
@@ -0,0 +1,596 @@
+}
+```
+
+### Capabilities
+
+The `Capability` type is extended with the new field `let id: UInt64`.
+
+### AuthAccount Type
+
+The capability management functions in `AuthAccount` will be grouped in nested objects,
+similar to how the contract management functions are in a nested [`contracts` object](https://developers.flow.com/cadence/language/accounts).
+
+Functions related to linking will be removed, as they are no longer needed.
+
+```cadence
+pub struct AuthAccount {
+    // ...
+    // removed: fun link<T: &Any>(_ newCapabilityPath: CapabilityPath, target: Path): Capability<T>?
+    // removed: fun unlink(_ path: CapabilityPath)
+    // removed: fun getLinkTarget(_ path: CapabilityPath): Path?
+    // moved and renamed:
+    //   old: fun getCapability<T: &Any>(_ path: CapabilityPath): Capability<T>
+    //   new: fun Capabilities.get<T: &Any>(_ path: CapabilityPath): Capability<T>?
+
+    pub let capabilities: &AuthAccount.Capabilities
+
+    pub struct Capabilities {
+
+        pub let storage: &AuthAccount.StorageCapabilities
+        pub let account: &AuthAccount.AccountCapabilities
+
+        /// Returns the capability at the given public path.
+        /// Returns nil if the capability does not exist,
+        /// or if the given type is not a supertype of the capability's borrow type.
+        pub fun get<T: &Any>(_ path: PublicPath): Capability<T>?
+
+        /// Borrows the capability at the given public path.
+        /// Returns nil if the capability does not exist, or cannot be borrowed using the given type.
+        /// The function is equivalent to `get(path)?.borrow()`.
+        pub fun borrow<T: &Any>(_ path: PublicPath): T?
+
+        /// Publish the capability at the given public path.
+        ///
+        /// If there is already a capability published under the given path, the program aborts.
+        ///
+        /// The path must be a public path, i.e., only the domain `public` is allowed.
+        pub fun publish(_ capability: Capability, at: PublicPath)
+
+        /// Unpublish the capability published at the given path.
+        ///
+        /// Returns the capability if one was published at the path.
+        /// Returns nil if no capability was published at the path.
+        pub fun unpublish(_ path: PublicPath): Capability?
+    }
+
+    pub struct StorageCapabilities {
+
+        /// Get the storage capability controller for the capability with the specified ID.
+        /// Returns nil if the ID does not reference an existing storage capability.
+        pub fun getController(byCapabilityID: UInt64): &StorageCapabilityController?
+
+        /// Get all storage capability controllers for capabilities that target this storage path
+        pub fun getControllers(forPath: StoragePath): [&StorageCapabilityController]
+
+        /// Iterate through all storage capability controllers for capabilities that target this storage path.
+        /// Returning false from the function stops the iteration.
```

Comment:
```suggestion

        /// Iterate over all storage capability controllers for capabilities that target this storage path,
        /// passing a reference to each controller to the provided callback function.
        ///
        /// Iteration is stopped early if the callback function returns `false`.
        ///
        /// If a new storage capability controller is issued for the path,
        /// an existing storage capability controller for the path is deleted,
        /// or a storage capability controller is retargeted from or to the path,
        /// then the callback must stop iteration by returning false.
        /// Otherwise, iteration aborts.
```